File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.py
def solution(s):
    flag = True
    n = len(s)
    left = []
    right = []
    # 처음에 )가 들어간 경우 = false
    if s[0]==')':
        return False
    
    for i in range(n):
        if s[i]=='(' and len(left)>=len(right):
            left.append(s[i])
        elif s[i]=='(' and len(left)<len(right):
            flag = False
            break
        elif s[i]==')' and len(left)>len(right):
            right.append(s[i])
        elif s[i]==')' and len(left)<=len(right):
            flag = False
            break
    
    if len(right)!=len(left):
        flag = False
        
        
    return flag

File: 백준/Gold/24552. 올바른 괄호/올바른 괄호.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String str = br.readLine();
        int left = 0;
        int right = 0;
        int sum = 0;
        for (int i = 0; i<str.length(); i++) {
            if (str.charAt(i)=='(') {
                left++;
                sum++;
            } else {
                right++;
                sum--;
            }
            if (sum<0) {
                System.out.println(right);
                return;
            }
            if (sum==0) {
                left = 0;
            }
        }
        System.out.println(left);
    }
}